### ISSUE TITLE
[OpenMP] Compilation for i686 fails with `error: unsupported size for integer register` when using GCC

### ISSUE THREAD

**Issue by giordano** [llvm/llvm-project#55006]
I'm trying to compile OpenMP 14.0.1 for `i686-linux-gnu` but I get the following error:
```
[ 51%] Building CXX object runtime/src/CMakeFiles/omp.dir/kmp_lock.cpp.o
cd /workspace/srcdir/openmp-14.0.1.src/build/runtime/src && /opt/bin/i686-linux-gnu-libgfortran5-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -Domp_EXPORTS -I/workspace/srcdir/openmp-14.0.1.src/build/runtime/src -I/workspace/srcdir/openmp-14.0.1.src/runtime/src -I/workspace/srcdir/openmp-14.0.1.src/runtime/src/i18n -I/workspace/srcdir/openmp-14.0.1.src/runtime/src/include -I/workspace/srcdir/openmp-14.0.1.src/runtime/src/thirdparty/ittnotify -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -std=c++14 -O3 -DNDEBUG -fPIC   -D _GNU_SOURCE -D _REENTRANT -fno-exceptions -fno-rtti -Wno-class-memaccess -Wno-covered-switch-default -Wno-frame-address -Wno-strict-aliasing -Wstringop-overflow=0 -Wno-stringop-truncation -Wno-switch -Wno-uninitialized -Wno-unused-but-set-variable -Wno-return-type-c-linkage -Wno-cast-qual -Wno-int-to-void-pointer-cast -msse2 -MD -MT runtime/src/CMakeFiles/omp.dir/kmp_lock.cpp.o -MF CMakeFiles/omp.dir/kmp_lock.cpp.o.d -o CMakeFiles/omp.dir/kmp_lock.cpp.o -c /workspace/srcdir/openmp-14.0.1.src/runtime/src/kmp_lock.cpp
/workspace/srcdir/openmp-14.0.1.src/runtime/src/kmp_lock.cpp: In function ‘void __kmp_spin_backoff(kmp_backoff_t*)’:
/workspace/srcdir/openmp-14.0.1.src/runtime/src/kmp_lock.cpp:2684:1: error: unsupported size for integer register
 }
 ^
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-int-to-void-pointer-cast’
cc1plus: warning: unrecognized command line option ‘-Wno-return-type-c-linkage’
cc1plus: warning: unrecognized command line option ‘-Wno-covered-switch-default’
```
```console
sandbox:${WORKSPACE}/srcdir/openmp-14.0.1.src/build # cc --version
i686-linux-gnu-gcc (GCC) 8.1.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

sandbox:${WORKSPACE}/srcdir/openmp-14.0.1.src/build # cc -dumpmachine
i686-linux-gnu
```
Host is `x86_64-linux-musl`.  Compilation is successful if using Clang 13.0.1.

**Comment by shiltian** [comment]
Could you please try the latest version? If the latest version doesn't work either, can you let us know how you built it?

**Comment by giordano** [comment]
I just tried to compile the current stable version of llvm-openmp, 15.0.6, for i686-linux-gnu, and build is successful with GCC from version 5 to version 12, so the bug I had reported with the then stable version seems to be fixed now :slightly_smiling_face: 

Side note, compiling llvm-openmp 16.06 with GCC 12 I get a couple of warnings:
```
[22:47:41] In file included from /workspace/srcdir/openmp-15.0.6.src/runtime/src/kmp.h:81,
[22:47:41]                  from /workspace/srcdir/openmp-15.0.6.src/runtime/src/kmp_runtime.cpp:13:
[22:47:41] /workspace/srcdir/openmp-15.0.6.src/runtime/src/kmp_runtime.cpp: In function ‘void* __kmp_launch_thread(kmp_info_t*)’:
[22:47:41] /workspace/srcdir/openmp-15.0.6.src/runtime/src/kmp_os.h:1141:26: warning: value computed is not used [-Wunused-value]
[22:47:41]  1141 | #define TCR_SYNC_PTR(a) ((void *)TCR_SYNC_4(a))
[22:47:41]       |                         ~^~~~~~~~~~~~~~~~~~~~~~
[22:47:41] /workspace/srcdir/openmp-15.0.6.src/runtime/src/kmp_runtime.cpp:6050:3: note: in expansion of macro ‘TCR_SYNC_PTR’
[22:47:41]  6050 |   TCR_SYNC_PTR((intptr_t)__kmp_global.g.g_done);
[22:47:41]       |   ^~~~~~~~~~~~
[22:47:41] At global scope:
[22:47:41] cc1plus: note: unrecognized command-line option ‘-Wno-int-to-void-pointer-cast’ may have been intended to silence earlier diagnostics
[22:47:41] cc1plus: note: unrecognized command-line option ‘-Wno-return-type-c-linkage’ may have been intended to silence earlier diagnostics
[22:47:41] cc1plus: note: unrecognized command-line option ‘-Wno-covered-switch-default’ may have been intended to silence earlier diagnostics
[...]
[22:47:57] /workspace/srcdir/openmp-15.0.6.src/libompd/src/omp-debug.cpp: In function ‘ompd_rc_t ompd_get_omp_version_string(ompd_address_space_handle_t*, const char**)’:
[22:47:57] /workspace/srcdir/openmp-15.0.6.src/libompd/src/omp-debug.cpp:145:27: warning: format ‘%ld’ expects argument of type ‘long int’, but argument 3 has type ‘ompd_word_t’ {aka ‘long long int’} [-Wformat=]
[22:47:57]   145 |   sprintf(omp_version, "%ld", ver);
[22:47:57]       |                         ~~^   ~~~
[22:47:57]       |                           |   |
[22:47:57]       |                           |   ompd_word_t {aka long long int}
[22:47:57]       |                           long int
[22:47:57]       |                         %lld
```

**Comment by shiltian** [comment]
Thanks for the note. Glad to see it's gone now. We will fix those warnings accordingly.
